R/PipeOpNearmiss.R

```diff
@@ -0,0 +1,109 @@
+#'
+#' @section Construction:
+#' ```
+#' PipeOpNearmiss$new(id = "nearmiss", param_vals = list())
+#' ```
+#'
+#' * `id` :: `character(1)`\cr
+#'   Identifier of resulting object, default `"nearmiss"`.
+#' * `param_vals` :: named `list`\cr
+#'   List of hyperparameter settings, overwriting the hyperparameter settings that would otherwise be set during construction. Default `list()`.
+#'
+#' @section Input and Output Channels:
+#' Input and output channels are inherited from [`PipeOpTaskPreproc`].
+#'
+#' The output during training is the input [`Task`][mlr3::Task] with the rows removed from the non-minority classes.
+#' The output during prediction is the unchanged input.
+#'
+#' @section State:
+#' The `$state` is a named `list` with the `$state` elements inherited from [`PipeOpTaskPreproc`].
+#'
+#' @section Parameters:
+#' The parameters are the parameters inherited from [`PipeOpTaskPreproc`], as well as
+#' * `k` :: `integer(1)`\cr
+#'   Number of nearest neighbors used for calculating the mean distances. Default is `5`.
+#' * `under_ratio` :: `numeric(1)`\cr
+#'   Ratio of the minority-to-majority frequencies. This specifies the ratio to which the number of instances
+#'   in the non-minority classes get down-sampled to, relative to the number of instances of the minority class.
+#'   Default is `1`. For details, see [`themis::nearmiss`].
+#'
+#' @section Fields:
+#' Only fields inherited from [`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @section Methods:
+#' Only methods inherited from [`PipeOpTaskPreproc`]/[`PipeOp`].
+#'
+#' @references
+#' `r format_bib("zhang2003")`
+#'
+#' @family PipeOps
+#' @template seealso_pipeopslist
+#' @include PipeOpTaskPreproc.R
+#' @export
+#' @examples
+#' \dontshow{ if (requireNamespace("themis")) \{ }
+#' library("mlr3")
+#'
+#' # Create example task
+#' task = tsk("wine")
+#' task$head()
+#' table(task$data(cols = "type"))
+#'
+#' # Down-sample and balance data
+#' pop = po("nearmiss")
+#' nearmiss_result = pop$train(list(task))[[1]]$data()
+#' nrow(nearmiss_result)
+#' table(nearmiss_result$type)
+#' \dontshow{ \} }
+PipeOpNearmiss = R6Class("PipeOpNearmiss",
+  inherit = PipeOpTaskPreproc,
+  public = list(
+    initialize = function(id = "nearmiss", param_vals = list()) {
+      ps = ps(
+        k = p_int(lower = 1, default = 5, tags = c("train", "nearmiss")),
+        under_ratio = p_dbl(lower = 0, default = 1, tags = c("train", "nearmiss"))
+      )
+      super$initialize(id, param_set = ps, param_vals = param_vals, packages = "themis", can_subset_cols = FALSE,
+                       task_type = "TaskClassif", tags = "imbalanced data")
+    }
+  ),
+  private = list(
+
+    .train_task = function(task) {
+      cols = task$feature_names
+
+      # Return task unchanged, if no feature columns exist
+      if (!length(cols)) {
+        return(task)
+      }
+      # Only numeric and integer features allowed
+      if (!all(task$feature_types$type %in% c("numeric", "integer"))) {
```

Comment:
only give error if there are no feature types of this kind; ignore non-numeric/integer cols instead